### PR TITLE
Modify softmax decomposition to be more numerically stable.

### PR DESCRIPTION
**Pull request by pashu123**
The softmax decomposition is modified according to https://github.com/pytorch/functorch/blob/main/functorch/_src/decompositions.pytorch
to account for numerical stability. Also, modified aten.argmax lowering
to handle negative dimension.

### CODE REVIEW

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 1]
```diff
@@ -148,22 +181,29 @@ class DecomposeAtenSelectIntOp : public OpRewritePattern<AtenSelectIntOp> {
...
   Value dim = op.dim();
   Value self = op.self();
-
-  // exp(x)
-  Value exp = rewriter.create<AtenExpOp>(loc, resultType, self);
-  // sum(exp(x))
-  Value sum =
-      createSumAlongDimension(rewriter, loc, op, exp, dim, /*keepDim=*/true);
+  Value xMax =
+      createMaxAlongDimension(rewriter, loc, op, self, dim, /*keepDim*/ true);
```
Comment (cathyzhyi) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:195]: nit: /*keepDim=*/true
Can you help to run `clang-format` on the entire file? Seems the format of quite a few places are off.

**Review by cathyzhyi — CHANGES_REQUESTED** [review]
looks mostly good. just 2 comments.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 1 above)
Comment (pashu123) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Done, thanks for noticing.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 2]
```diff
@@ -33,17 +33,15 @@ static int getTensorRank(Value tensor) {
...
+                                 Value input, Value dim, bool keepDim) {
   BaseTensorType tensorType = input.getType().cast<BaseTensorType>();
-  Value dimList = rewriter.create<PrimListConstructOp>(
-      loc, Torch::ListType::get(dim.getType()), dim);
-  Value keepDimCst = rewriter.create<ConstantBoolOp>(loc, keepDim);
-  Value dtype = rewriter.create<ConstantNoneOp>(loc);
   SmallVector<int64_t> sizes;
   int64_t dimInt;
-  if (tensorType.hasSizes()) {
+  if (tensorType.hasSizes() && keepDim) {
```
[the remark this answers is not in the document]
Comment (pashu123) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Done, thanks.

**Review by cathyzhyi — APPROVED** [review]
LGTM. Thanks!
